File: Python/3_LongestSubstringWithoutRepeatingCharacters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        def bt(s,idx,num,ts):   # ts is temporal strings
            if idx == len(s):   # base
                return num
            else:
                if s[idx] not in ts:
                    return bt(s,idx+1,num+1,ts+s[idx])
                else:
                    return num
            
        mres = 0

        for i in range(len(s)):
            mres = max(mres,bt(s,i,0,""))  
            
        return mres

# backtracking and memoization

File: C++/3_LongestSubstringWithoutRepeatingCharacters.cpp
using namespace std;
#include <vector>

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<bool> ht(256,false);

        int maxCount=0;
        int l=0;
        int r=0;

        for (auto c:s) {
            while(ht[c]){
                ht[s[l++]]=false;
            }
            ht[c]=true;

            maxCount = max(maxCount, r-l+1);
            r++;
        }

        return maxCount;
        
    }
};